url835a90ed59_aschi_regularLanguages_tgz-pJ8-ch_zhaw_regularLanguages_evolution_runners_EvolvingLocalDivisable5LanguageJ8/src/fr/loria/graphViz/GraphViz.java

```diff
@@ -0,0 +1,288 @@
+    * @return The image of the graph in .gif format.
+    */
+   private byte[] get_img_stream(File dot, String type)
+   {
+      File img;
+      byte[] img_stream = null;
+
+      try {
+         img = File.createTempFile("graph_", "."+type, new File(GraphViz.TEMP_DIR));
+         Runtime rt = Runtime.getRuntime();
+         
+         // patch by Mike Chenault
+         String[] args = {DOT, "-T"+type, dot.getAbsolutePath(), "-o", img.getAbsolutePath()};
+         Process p = rt.exec(args);
+         
+         p.waitFor();
+
+         FileInputStream in = new FileInputStream(img.getAbsolutePath());
+         img_stream = new byte[in.available()];
+         in.read(img_stream);
+         // Close it if we need to
+         if( in != null ) in.close();
+
+         if (img.delete() == false) 
+            System.err.println("Warning: " + img.getAbsolutePath() + " could not be deleted!");
+      }
+      catch (java.io.IOException ioe) {
+         System.err.println("Error:    in I/O processing of tempfile in dir " + GraphViz.TEMP_DIR+"\n");
+         System.err.println("       or in calling external command");
+         ioe.printStackTrace();
+      }
+      catch (java.lang.InterruptedException ie) {
+         System.err.println("Error: the execution of the external program was interrupted");
+         ie.printStackTrace();
+      }
+
+      return img_stream;
+   }
+
+   /**
+    * Writes the source of the graph in a file, and returns the written file
+    * as a File object.
+    * @param str Source of the graph (in dot language).
+    * @return The file (as a File object) that contains the source of the graph.
+    */
+   private File writeDotSourceToFile(String str) throws java.io.IOException
+   {
+      File temp;
+      try {
+         temp = File.createTempFile("graph_", ".dot.tmp", new File(GraphViz.TEMP_DIR));
+         FileWriter fout = new FileWriter(temp);
+         fout.write(str);
+         fout.close();
+      }
+      catch (Exception e) {
+         System.err.println("Error: I/O error while writing the dot source to temp file!");
+         return null;
+      }
+      return temp;
+   }
+
+   /**
+    * Returns a string that is used to start a graph.
+    * @return A string to open a graph.
+    */
+   public String start_graph() {
+      return "digraph G {";
+   }
+
+   /**
+    * Returns a string that is used to end a graph.
+    * @return A string to close a graph.
+    */
+   public String end_graph() {
+      return "}";
+   }
+
+   /**
+    * Read a DOT graph from a text file.
+    * 
+    * @param input Input text file containing the DOT graph
+    * source.
+    */
+   public void readSource(String input)
+   {
+	   StringBuilder sb = new StringBuilder();
+	   
+	   try
+	   {
+		   FileInputStream fis = new FileInputStream(input);
+		   DataInputStream dis = new DataInputStream(fis);
+		   BufferedReader br = new BufferedReader(new InputStreamReader(dis));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *br*. The resource is referenced at lines 272, 273, 275. The resource is closed at line 278. There are other execution paths that do not contain closure statements, for example, when BufferedReader.readLine throws an exception. Ensure *br* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).